### ISSUE TITLE
Require inside of a client.d file does not work on first run

### ISSUE THREAD

**Issue by martinisoft** [chef/chef#7088]
## Description

When you attempt to load a custom gem via the chef `client.d` folder in Chef Client (Chef Zero is also affected) it will not be available and raise an exception at the end of the run in which it is created and loaded. Subsequent runs will work normally.

## Chef Version

Chef 12 and 13

## Platform Version

Tested platforms:
Ubuntu 14.04 & 16.04
CentOS 6 & 7

## Replication Case

I have created [a pull request](https://github.com/chef-cookbooks/chef-client/pull/567) which uses the example in the README (with corrections) in the test suite to demonstrate the problem. My hunch is that `Chef::Config[:client_d_dir]` for the chef_zero provisioner is going to `/tmp` and not loading the gem path to the newly installed gem via metadata.rb or the chef_gem directive. Somehow it's still being called at the end of the chef run.

## Client Output

```
I, [2018-03-29T00:17:44.656488 #5478]  INFO -- default-ubuntu-1604: Running handlers:
I, [2018-03-29T00:17:44.656630 #5478]  INFO -- default-ubuntu-1604: [2018-03-29T05:17:42+00:00] INFO: Running report handlers
I, [2018-03-29T00:17:44.656784 #5478]  INFO -- default-ubuntu-1604: Running handlers complete
I, [2018-03-29T00:17:44.657008 #5478]  INFO -- default-ubuntu-1604: [2018-03-29T05:17:42+00:00] INFO: Report handlers complete
I, [2018-03-29T00:17:44.657148 #5478]  INFO -- default-ubuntu-1604: Chef Client finished, 88/140 resources updated in 02 minutes 03 seconds
I, [2018-03-29T00:17:44.657290 #5478]  INFO -- default-ubuntu-1604: [2018-03-29T05:17:42+00:00] DEBUG: Server doesn't support resource history, skipping resource report.
I, [2018-03-29T00:17:44.657439 #5478]  INFO -- default-ubuntu-1604: [2018-03-29T05:17:42+00:00] DEBUG: Audit Reports are disabled. Skipping sending reports.
I, [2018-03-29T00:17:44.657576 #5478]  INFO -- default-ubuntu-1604: [2018-03-29T05:17:42+00:00] DEBUG: Forked instance successfully reaped (pid: 1562)
I, [2018-03-29T00:17:44.657719 #5478]  INFO -- default-ubuntu-1604: [2018-03-29T05:17:42+00:00] FATAL: Configuration error LoadError: cannot load such file -- bugsnag
I, [2018-03-29T00:17:44.657899 #5478]  INFO -- default-ubuntu-1604: [2018-03-29T05:17:42+00:00] FATAL:   /tmp/kitchen/client.d/bugsnag.rb:1:in `from_string'
I, [2018-03-29T00:17:44.658042 #5478]  INFO -- default-ubuntu-1604: [2018-03-29T05:17:42+00:00] FATAL: Aborting due to error in '/tmp/kitchen/client.d/bugsnag.rb'
```

## Stacktrace

```
I, [2018-03-29T00:17:44.657719 #5478]  INFO -- default-ubuntu-1604: [2018-03-29T05:17:42+00:00] FATAL: Configuration error LoadError: cannot load such file -- bugsnag
I, [2018-03-29T00:17:44.657899 #5478]  INFO -- default-ubuntu-1604: [2018-03-29T05:17:42+00:00] FATAL:   /tmp/kitchen/client.d/bugsnag.rb:1:in `from_string'
I, [2018-03-29T00:17:44.658042 #5478]  INFO -- default-ubuntu-1604: [2018-03-29T05:17:42+00:00] FATAL: Aborting due to error in '/tmp/kitchen/client.d/bugsnag.rb'
```

## Workaround

For the time being you can force the Gem Paths to reload via `Gem.clear_paths` before your gem and require call:

```ruby
Gem.clear_paths
gem 'bugsnag'
require 'bugsnag'
Bugsnag.configure do |config|
  config.api_key = '1234567890'
end
```

**Comment by martinisoft** [comment]
I have now confirmed this also affects Chef-Client 13.8 without Chef Zero mode so this is a core chef problem.

**Comment by lamont-granquist** [comment]
So earlier in the run you've successfully done `chef_gem "bugsnag"` or installed it via metadata.rb then?

Offhand, I can only be puzzled.  It smells like bundler, but we don't bundle inside of TK AFAIK, and AFAIK we don't do anything weird with rubygems in report handlers, and would probably look more in the direction of the rubygems package resource behaving strangely.  If its a regression it might be due to the recent patch there to work around the embedded-gems problem.

**Comment by martinisoft** [comment]
> So earlier in the run you've successfully done chef_gem "bugsnag" or installed it via metadata.rb then?

Correct. I had a cookbook in the run list as the first item which drops off this `client.d` file and in it's metadata is a `gem 'bugsnag'` so at the beginning of the run it installs. Near the end of the run the `client.d` file is dropped off. Somewhere between there and the end of the run the file gets loaded and thus crashes, but without any stacktrace.

**Comment by lamont-granquist** [comment]
```
 DEBUG: Forked instance successfully reaped (pid: 1562)
```

that line indicates that the forked process was successfully reaped and is gone.

what is happening is that the daemon process re-reads the client.d file now and the `require "bugsnag"` causes it to blow up even though the gem was installed by the child process, which smells a lot like its the internal gem state of the parent process, and likely `Gem.clear_paths` put on the line before the `require` will fix it.

**Comment by martinisoft** [comment]
> what is happening is that the daemon process re-reads the client.d file now and the require "bugsnag" causes it to blow up even though the gem was installed by the child process, which smells a lot like its the internal gem state of the parent process, and likely Gem.clear_paths put on the line before the require will fix it.

And indeed that did fix the issue. I have updated my original post about the issue for others to find as a solution.

**Comment by martinisoft** [comment]
This is now resolved by #7106 for Chef 14. Anyone on 12 and 13 can do the workaround to resolve this issue. Thank you for the assistance @lamont-granquist and @coderanger ❤️
